### PR TITLE
Added more examples on Join types

### PR DESCRIPTION
**Pull request by shnapz**
- More examples on join types to enrich our public docs
- Small code simplification in skewed implementation

### CODE REVIEW

File: scio-examples/src/main/scala/com/spotify/scio/examples/cookbook/JoinExamples.scala [hunk 1]
```diff
@@ -144,3 +144,74 @@ object HashJoinExamples {
...
+    eventsInfo
+      // Skewed join is useful when LHS contains a subset of keys with high frequency, but RHS is
+      // too large to fit into memory. It uses Count Min Sketch to estimate those frequencies.
+      // Internally it identifies two groups of keys: "Hot" and the rest. Hot keys are joined using
+      // Hash Join and the rest with the regular join. There are 3 ways to identify
+      // the set of hot keys:
+      // 1) "threshold" as a cutoff frequency
+      // 2) "top percentage" to specify the maximum relative part of all keys can be considered hot
+      // 3) "top N" to specify the absolute number of hot keys
+      .skewedLeftOuterJoin(countryInfo, hotKeyThreshold = 100)
```
Comment (RustedBones) [inline scio-examples/src/main/scala/com/spotify/scio/examples/cookbook/JoinExamples.scala:170]: That's nice.
Should we go a bit further and use the new API, with some explanation on how to pick good params for the CMS (epsilon, delta)?

Comment (shnapz) [reply scio-examples/src/main/scala/com/spotify/scio/examples/cookbook/JoinExamples.scala:170]: Agree!

File: scio-examples/src/main/scala/com/spotify/scio/examples/cookbook/JoinExamples.scala [hunk 2]
```diff
@@ -144,3 +144,87 @@ object HashJoinExamples {
...
+      // than 1.0, it may be also required to fit keys sample in memory for CMS. If you sample only
+      // 0.1 of the dataset then you need to decrease "threshold" 10 times respectively, because the
+      // latter relies on absolute frequencies detected in the sample.
+      // - "withReplacement" - if "true" it will use Poisson distribution, otherwise Bernoulli. The
+      // former will allow repeats of the same item in your sample.
+      // - "hotKeyFanout" - tune Apache Beam fanout feature when aggregating sample keys to CMS
+      // vectors. It should be a positive number and specifies intermediate nodes to redistribute
+      // aggregation over heavy-hitter keys. Tune it when "Compute CMS of LHS keys" transform has
+      // a problem of idle workers.
+      // - "cmsEps", "cmsDelta", "cmsSeed" - are inputs to CMS probabilistic computation.
```
Comment (shnapz) [inline scio-examples/src/main/scala/com/spotify/scio/examples/cookbook/JoinExamples.scala:182]: @RustedBones  added more about params, but it's worth to expand on Epsilon and Delta

File: scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala [hunk 3]
```diff
@@ -60,13 +60,12 @@ object SkewedJoins {
...
   // some sensible defaults for skewed joins
   val DefaultHotKeyThreshold: Int = 9000
-  val DefaultHotKeyMethod: HotKeyMethod.Threshold = HotKeyMethod.Threshold(DefaultHotKeyThreshold)
   val DefaultHotKeyFanout: Int = 1
   val DefaultCmsEpsilon: Double = 0.001
   val DefaultCmsDelta: Double = 1e-10
   val DefaultCmsSeed: Int = 42
   val DefaultSampleFraction: Double = 1.0
-  val DefaultSampleWithReplacement: Boolean = true
+  val DefaultSampleWithReplacement: Boolean = false
```
Comment (shnapz) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala:68]: @RustedBones  why do we use Poisson as default sampling distribution? It introduces repeats of the same items

File: scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala [hunk 4]
```diff
@@ -60,13 +60,12 @@ object SkewedJoins {
 
   // some sensible defaults for skewed joins
   val DefaultHotKeyThreshold: Int = 9000
-  val DefaultHotKeyMethod: HotKeyMethod.Threshold = HotKeyMethod.Threshold(DefaultHotKeyThreshold)
```
Comment (shnapz) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala:63]: not used anywhere

Comment (RustedBones) [reply scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala:63]: Was left here because people ca use those default when calling the new API (as overloading with default is not allowed). The default constants we not made private.

File: scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala (hunk 3 above)
Comment (RustedBones) [reply scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala:68]: I just kept previous default to prevent change of behavior.

File: scio-examples/src/main/scala/com/spotify/scio/examples/extra/ParquetExample.scala [hunk 5]
```diff
@@ -136,9 +137,15 @@ object ParquetExample {
...
 
-  private def typedIn(sc: ScioContext, args: Args): ClosedTap[String] =
-    sc.typedParquetFile[AccountProjection](args("input"))
+  private def typedIn(sc: ScioContext, args: Args): ClosedTap[String] = {
+    // typed API does not support macro generated expressions, so it requires using FilterApi
+    val predicateExp = FilterApi.and(
+      FilterApi.notEq(FilterApi.intColumn("id"), java.lang.Integer.valueOf(0)),
+      FilterApi.lt(FilterApi.intColumn("id"), java.lang.Integer.valueOf(10000))
+    )
+    sc.typedParquetFile[AccountProjection](args("input"), predicate = predicateExp)
```
Comment (RustedBones) [inline scio-examples/src/main/scala/com/spotify/scio/examples/extra/ParquetExample.scala:146]: can we have this in another PR as this is not related ?

Comment (shnapz) [reply scio-examples/src/main/scala/com/spotify/scio/examples/extra/ParquetExample.scala:146]: Sure!

File: scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala (hunk 4 above)
Comment (shnapz) [reply scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala:63]: Ok

File: scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala (hunk 3 above)
Comment (shnapz) [reply scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala:68]: Let's change it in 0.13?

File: scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala [hunk 6]
```diff
@@ -161,11 +161,11 @@ class PairSkewedSCollectionFunctions[K, V](val self: SCollection[(K, V)]) {
    *   One-sided error bound on the error of each point query, i.e. frequency estimate. Must lie in
    *   `(0, 1)`.
    * @param cmsDelta
-   *   A seed to initialize the random number generator used to create the pairwise independent hash
-   *   functions.
-   * @param cmsSeed
    *   A bound on the probability that a query estimate does not lie within some small interval (an
    *   interval that depends on `eps`) around the truth. Must lie in `(0, 1)`.
+   * @param cmsSeed
```
Comment (clairemcginty) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSkewedSCollectionFunctions.scala:166]: nice catch!

File: scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala [hunk 7]
```diff
@@ -902,7 +902,10 @@ sealed trait SCollection[T] extends PCollectionWrapper[T] {
     this.pApply(Combine.globally(Functions.reduceFn(context, op)).withoutDefaults())
 
   /**
-   * Return a sampled subset of this SCollection.
+   * Return a sampled subset of this SCollection with the exact number of items. Involves combine
```
Comment (clairemcginty) [inline scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala:905]:
```suggestion
   * Return a sampled subset of this SCollection containing exactly `sampleSize` items. Involves combine
```
